### PR TITLE
draft ttls in marblerun

### PR DESCRIPTION
**Pull request by 3u13r**
The first draft to describe the TTLS feature. I tried to be as precise and short as possible. 
I saw that the other feature descriptions also explained the general use-case inside marblerun. This part probably needs to be more flashed out. 

I welcome all suggestions and corrections. Concrete questions are included below.

### CODE REVIEW

File: content/docs/features/transparent-TLS.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+---
+title: "Transparent TLS"
+date: 2020-11-14T16:28:16+05:30
+draft: false
+weight: 3
+---
+
+# Transparent TLS
+
+To achieve security and verifiability, services need to communicate via TLS. To simplify migrating existing applications to Marblerun, Transparent TLS (TTLS) can wrap any defined connections in TLS. Those connections can be defined in the manifest.
```
Comment (3u13r) [inline content/docs/features/transparent-TLS.md:10]: How much should we add here about why we use TLS in the first place. If I understand correctly this is mostly covered by features/secret-management and features/attestation. Should we move some parts here, reiterate the important parts (which?), or just reference it?

File: content/docs/workflows/define-manifest.md [hunk 2]
```diff
@@ -290,6 +297,42 @@ Use the following command to preserve newlines correctly:
 awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' public_key.pem
 ```
 
+
+## Manifest:TLS
+
+The TLS entry holds a list of tags which can be used in a marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. Outgoing connections are defined by `Port` and `Addr`. Incoming connections just use `Port` and always define `Addr` as `"*"`. Both domains and IPs can be used as `Addr`. Previously defined certificates can be referenced by name. This disables client authentication.
```
Comment (3u13r) [inline content/docs/workflows/define-manifest.md:303]: Should we warn about the disabling of client authentication in a highlighted box?

File: content/docs/workflows/define-manifest.md [hunk 3]
```diff
@@ -85,7 +88,10 @@ Marbles represent the actual services in your mesh. They are defined in the `Mar
                     "MARBLE_CERT": "{{ pem .Marblerun.MarbleCert.Cert }}",
                     "MARBLE_KEY": "{{ pem .Marblerun.MarbleCert.Private }}"
                 }
-            }
+            },
+            "TLS": [
+                "front", "other"
```
Comment (m1ghtym0) [inline content/docs/workflows/define-manifest.md:93]: Why don't we name them similar to the service?

File: content/docs/workflows/define-manifest.md [hunk 4]
```diff
@@ -290,6 +297,42 @@ Use the following command to preserve newlines correctly:
 awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' public_key.pem
 ```
 
+
+## Manifest:TLS
+
+The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. Outgoing connections are defined by `Port` and `Addr`. Incoming connections just use `Port` and always define `Addr` as `"*"`. Both domains and IPs can be used as `Addr`. Previously defined certificates can be referenced by name. This disables client authentication.
```
Comment (m1ghtym0) [inline content/docs/workflows/define-manifest.md:303]: Why do previously defined certificates disable client authentication?
Wouldn't it be better to have an explicit parameter for that?
`domains` are a bit ambiguous and probably require one more sentence on DNS names.

File: content/docs/workflows/define-manifest.md (hunk 3 above)
Comment (3u13r) [reply content/docs/workflows/define-manifest.md:93]: If I understand correctly service meaning marble?
I wanted to emphasize that the names of the tags are arbitrary and don't need to exactly match the services.

File: content/docs/workflows/define-manifest.md (hunk 4 above)
Comment (3u13r) [reply content/docs/workflows/define-manifest.md:303]: Currently, client authentication is automatically disabled as specified in https://github.com/edgelesssys/emojivoto/pull/11#issuecomment-790501876. This would only need a small change in marblerun to be able to configure it manually. There are probably use cases where client authentication can/must be enabled. If there are no objections, I will add this feature. 

I'll also add some more information on the `domains` part, here. Especially that the internal kubernetes domains can be used.

File: content/docs/features/transparent-TLS.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+date: 2020-11-14T16:28:16+05:30
+draft: false
+weight: 3
+---
+
+# Transparent TLS
+
+To achieve security and verifiability, services need to communicate via TLS. To simplify migrating existing applications to Marblerun, Transparent TLS (TTLS) can wrap any connection in TLS without any code changes. Those connections can be defined in the Manifest.
+
+While user-defined TLS credentials can be used, Marblerun defaults to the [marble's credentials]({{< ref "docs/features/secrets-management.md" >}}). Since all Marble's certificates are signed by the Coordinator's RootCA, connections between two Marbles are authenticated. For external connections, client authentication can be disabled.
```
[the remark this answers is not in the document]
Comment (katexochen) [reply content/docs/features/transparent-TLS.md:12]: "RootCA" -> "root CA"

File: content/docs/workflows/define-manifest.md (hunk 3 above)
Comment (daniel-weisse) [reply content/docs/workflows/define-manifest.md:93]: Maybe change the labels from `front` and `back` to `frontend_tls` and `backend_first_tls`?
This way we keep the naming from the service, but also show that the names don't need to match

File: content/docs/features/transparent-TLS.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+---
+title: "Transparent TLS"
+date: 2020-11-14T16:28:16+05:30
+draft: false
+weight: 3
+---
+
+# Transparent TLS
+
+To achieve security and verifiability, services need to communicate via TLS. To simplify migrating existing applications to Marblerun, Transparent TLS (TTLS) can wrap any connection in TLS without any code changes. Those connections can be defined in the Manifest.
```
[the remark this answers is not in the document]
Comment (thomasten) [reply content/docs/features/transparent-TLS.md]: After reading this, I wouldn't know why I'd need this feature. Should we start with a problem description?

File: content/docs/workflows/define-manifest.md (hunk 3 above)
Comment (thomasten) [reply content/docs/workflows/define-manifest.md]: TLS tags/labels should be named after what they provide, not after the marble(s) that use them.

File: content/docs/features/transparent-TLS.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (thomasten) [reply content/docs/features/transparent-TLS.md]: Start with the default.
The internal credentials were a crutch and are now an implementation detail - scratch that. Just point out that they are automatically configured.
Explain why/when you would define your own credentials.

File: content/docs/workflows/define-manifest.md [hunk 7]
```diff
@@ -290,6 +297,43 @@ Use the following command to preserve newlines correctly:
...
+                    "Addr": "127.0.0.1"
+                }
+            ],
+            "Incoming": [
+                {
+                    "Port": "8080",
+                    "Addr": "*",
+                    "Cert": "rsa_cert",
+                    "ClientAuth": false
+                }
```
Comment (thomasten) [inline content/docs/workflows/define-manifest.md:327]: We should also show how to define an incoming connection for the most common case, i.e., internal mutual TLS.

File: content/docs/workflows/define-manifest.md [hunk 8]
```diff
@@ -290,6 +297,43 @@ Use the following command to preserve newlines correctly:
...
+        "frontend_tls_1": {
+            "Outgoing": [
+                {
+                    "Port": "8080",
+                    "Addr": "service.name"
+                },
+                {
+                    "Port": "4443",
+                    "Addr": "127.0.0.1"
+                }
```
Comment (thomasten) [inline content/docs/workflows/define-manifest.md:319]: I think this is a bad example in the CC context.

File: content/docs/workflows/define-manifest.md [hunk 9]
```diff
@@ -290,6 +297,43 @@ Use the following command to preserve newlines correctly:
 awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' public_key.pem
 ```
 
+
+## Manifest:TLS
+
+The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. Outgoing connections are defined by `Port` and `Addr`. Incoming connections just use `Port` and always define `Addr` as `"*"`. Both domains and IPs can be used as `Addr`. Most importantly, the DNS names of other services can be used.  Previously defined certificates can be referenced by name. This disables client authentication.
```
Comment (thomasten) [inline content/docs/workflows/define-manifest.md:303]: > Incoming connections just use `Port` and always define `Addr` as `"*"`.

So why do they define `Addr` at all?

Comment (thomasten) [inline content/docs/workflows/define-manifest.md:303]: This paragraph should be split in two or three
* Short intro and common facts (if any)
* Incoming connections: Why, when, and how to use them (internal mutual vs external)
* Outgoing connections: Why, when, and how to use them

File: content/docs/workflows/define-manifest.md (hunk 3 above)
Comment (3u13r) [reply content/docs/workflows/define-manifest.md]: I'm kind of agree and disagree. IMO, the most probable scenario is something like 5 gRPC Services and maybe 1 web service. Now we maybe want to share the configuration of the gRPC services.  Even if they all listen on the same port, and the Incoming definition can be shared, we would still need to define each service in the Outgoing section. Therefore, we also define that Service1 connection to Service1 should be wrapped, which makes little sense. I see that it is nice to have all the services that talk to each other in one list (e.g. the Outgoing part), but this overdefinition can quickly explode, e.g. when Service1 only talks to Service2 and Service2 only to Service3. Seeing those services grouped as one TLS definition makes me think that Service1 talks to every service in the list.  Again, assuming they all use the same incoming port. Therefore, my way of configuring would be marble-wise, if there is no good reason not to. Maybe I would even define an "gRPC_incoming" tag, use that for each marble and define the Outgoing part separately for each.  

Furthermore, it is quite hard to show the different variations with 2 of them. 

tl;dr: 2 Options: By default TLS labels define 
1. marbles (1-to-1)  
2. "groups" of marbles (1-to-n)

File: content/docs/workflows/define-manifest.md (hunk 9 above)
Comment (3u13r) [reply content/docs/workflows/define-manifest.md:303]: We could leave it out and define it in marblerun ourselves. We also could leave it there to remind the user that we don't bind on a specific interface but all interfaces.

File: content/docs/workflows/define-manifest.md (hunk 3 above)
Comment (thomasten) [reply content/docs/workflows/define-manifest.md]: I don't get all of what you say, but doesn't matter.

> we would still need to define each service in the Outgoing section

You can have an Outgoing rule with a wildcard address and fixed port. One could argue that a wildcard address is always preferable regarding security. At least it's not dangerous.

> Maybe I would even define an "gRPC_incoming" tag

Yes, that's my point.

In general, the feature is designed to support n-to-m. (One reason is avoiding duplicate entries in multiple TLS definititions.)

Anyway, I didn't necessarily want anything to be changed in this example, but just clarify to all participants that "name them similar to the service" is not the rule of thumb.

File: content/docs/workflows/define-manifest.md (hunk 9 above)
Comment (thomasten) [reply content/docs/workflows/define-manifest.md:303]: Sorry, I thought we already support leaving it out. So is the following valid or not?
https://github.com/edgelesssys/marblerun/blob/3891ebcbf717e51218c9c3bb3b66270f6dd9713b/test/manifests.go#L152-L155

Comment (3u13r) [reply content/docs/workflows/define-manifest.md:303]: Sorry, I had to look into the marblerun implementation again. Currently, your example is already supported. I have no idea why I defined Addr as "*". So I will just remove the sentence and replace it with something like: "Incoming connections just use `Port` and which will be used on all interfaces."

**Comment by 3u13r** [comment]
Just for visibility: Now added (in https://github.com/edgelesssys/marblerun.sh/pull/49/commits/9ec7577eb86ccf1b91bbb413344d63b11a49e8ef) the change requests by @thomasten as well as the `DisableClientAuth` flag.

File: content/docs/workflows/define-manifest.md [hunk 10]
```diff
@@ -290,6 +297,50 @@ Use the following command to preserve newlines correctly:
...
 ```
 
+
+## Manifest:TLS
+
+The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. To elevate the connection between two marbles to TLS, the client needs to set the server under `Incoming` and the server needs to define its service under `Incoming`.
+
+Outgoing connections are defined by `Port` and `Addr`. Both domains and IPs can be used as `Addr`. Most importantly, the DNS names of other services can be used.
+
+Incoming connections are listened to on all interfaces and are only defined by `Port`. For services that are used by external clients, you must set `Cert`. Previously defined certificates can be referenced here by name. Client authentication then needs to be disabled by setting `DisableClientAuth` to `true`.
```
Comment (thomasten) [inline content/docs/workflows/define-manifest.md:307]:
```suggestion
Incoming connections are defined by `Port`. All interfaces will be intercepted.. For services that are used by external clients, you must set `Cert`. Previously defined certificates can be referenced here by name. Client authentication then needs to be disabled by setting `DisableClientAuth` to `true`.
```
"Incoming connections are listened to on all interfaces" reads odd. Maybe even wrong as it may suggest that TTLS controls on what interfaces to listen. However, my suggestion still is hard to understand without more context. Does it even add value to mention this at all? May be more confusing than helpful to most readers.

Comment (thomasten) [inline content/docs/workflows/define-manifest.md:307]:
```suggestion
Incoming connections are listened to on all interfaces and are only defined by `Port`. For services used by external clients, you must disable client authentication by setting `DisableClientAuth` to `true` and set `Cert`. Use the name of a certificate defined in the [Secrets section](#manifestsecrets).
```
I tried to make the reason for disabling client auth more clear by (grammatically) binding it to "external clients".

File: content/docs/workflows/define-manifest.md [hunk 11]
```diff
@@ -290,6 +297,50 @@ Use the following command to preserve newlines correctly:
 awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' public_key.pem
 ```
 
+
+## Manifest:TLS
+
+The TLS entry holds a list of tags which can be used in a Marble's definition. Each tag can define multiple `Incoming` and `Outgoing` connections. To elevate the connection between two marbles to TLS, the client needs to set the server under `Incoming` and the server needs to define its service under `Incoming`.
+
+Outgoing connections are defined by `Port` and `Addr`. Both domains and IPs can be used as `Addr`. Most importantly, the DNS names of other services can be used.
```
Comment (thomasten) [inline content/docs/workflows/define-manifest.md:305]:
```suggestion
Outgoing connections are defined by `Port` and `Addr`. For `Addr`, you can use both IP addresses and domains, e.g., the DNS names of other services.
```
prefer active voice where possible (if it doesn't complicate the sentence or sound odd)

**Review by thomasten — CHANGES_REQUESTED** [review]

File: content/docs/workflows/add-service.md [hunk 12]
```diff
@@ -11,13 +11,16 @@ Adding a service to your application requires three steps, which are described i
 
 ## **Step 1:** Get your service ready for Marblerun
 
-To get your service ready for Marblerun, you need to rebuild it with one of the supported [runtimes]({{< ref "docs/features/runtimes.md" >}}):
-* [EGo]({{< ref "docs/building-services/ego.md" >}})
+We recommend using [EGo]({{< ref "docs/building-services/ego.md" >}}) as runtime for all Go applictions, since there is no recompiling necessary when using [TTLS]({{< ref "docs/features/transparent-TLS.md" >}}).
```
Comment (thomasten) [inline content/docs/workflows/add-service.md:14]: You still need to recompile with EGo. I think you can just revert this change.

**Comment by thomasten** [comment]
Can one of the other reviewers (@katexochen @daniel-weisse @m1ghtym0) review this again pls?

**Review by m1ghtym0 — APPROVED** [review]
LGTM:-)

**Review by thomasten — APPROVED** [review]
